content/en/blog/2023/2023-03-08-scheduler-plugin-reusing/index.md

```diff
@@ -0,0 +1,124 @@
+
+Going through the exercise of building a new plugin via composition and realizing the caveats of such reuse raised some questions related to if and how things could be better in an ideal world. Such as:
+
+* Could a plugin utilize scores from multiple other plugins?
+* Could multiple new plugins utilize scores from the same plugin, without running the same scoring logic N times?
+* Could the dependency to the reused plugin be made less fragile, reducing the coupling?
+* Could reuse be achieved without duplicated args and configuring same algorithms the same way, twice (or more)?
+* Could things be configured more dynamically instead of hard-coding to the sources one plugin uses another plugin?
+
+These questions had the author thinking about finding a better solution than just composition. Also the documentation of the [framework.CycleState](https://github.com/kubernetes/kubernetes/blob/2f977fd8c400971deed365dcc437e519bce475f0/pkg/scheduler/framework/cycle_state.go#L42-L47) was looking rather tempting.
+
+> **Warning**
+> The following section is speculative and based on proof-of-concepting by the author. It doesn't represent what you can do with the Kubernetes scheduler at the time of writing. Some day the concept below might be part of Kubernetes, should the PoC develop into a successful KEP. For now, this is merely a scheduling-sig discussion item which shows the potential of the CycleState.
+
+Ideally, a new plugin could ask for getting the scores from another plugin, or multiple plugins, without rerunning them. The tricky thing is, that plugins run highly parallel in order to execute faster. Thus proper synchronization would be needed. Also it would be ideal, if no changes to existing plugins would be required, and the framework would provide for plugins in need of reusing other plugins' results.
+
+This got the author thinking, that if the `framework.CycleState` object could contain a channel for the scores, and if the framework would create such a channel when it is needed, then the parallel running plugins which are interested in score reuse, could wait for the scores to appear from such "ScoreChannels".
+
+```go
+// ScoreChannel for giving score results to other plugins.
+type ScoreChannel struct {
+	C chan framework.NodeScoreList
+}
+```
+
+A little less than 80 framework lines of PoC-code later the now very dirty scheduler was able to pick up from the arguments of plugins, whether they wanted to use scores from other plugins, and it was creating such "ScoreChannels" into the `framework.CycleState` and wrote scores to them as requested. A couple of experimental plugins later it was possible to verify that indeed, plugin scoring graphs such as the following were entirely possible:
+
+{{< figure src="diagram1.svg" alt="Plugins A, B, C and D with connections A->B, A->C, B->D, C->D">}}
+
+Which makes a nice demo of the one-to-many and many-to-one pipelining possibilities for plugin score reusing.
+
```

Comment:
could you put a pseudo code snippet here to explain how TopologyScorer leverages the score it get?

Reply:
Ok, will do.